aws/rust-runtime/aws-types/src/sdk_config.rs

```diff
@@ -499,6 +502,105 @@ impl Builder {
+    ///         let user_agent = AwsUserAgent::for_tests();
+    ///         headers.insert(USER_AGENT, HeaderValue::try_from(user_agent.ua_header())?);
+    ///         headers.insert(
+    ///             HeaderName::from_static("x-amz-user-agent"),
+    ///             HeaderValue::try_from(user_agent.aws_ua_header())?,
+    ///         );
+    ///
+    ///         Ok(())
+    ///     }
+    /// }
+    ///
+    /// let sdk_config = SdkConfig::builder()
+    ///     .interceptor(TestUserAgentInterceptor)
+    ///     .build();
+    /// # }
+    /// ```
+    pub fn interceptor(mut self, interceptor: impl Interceptor + Send + Sync + 'static) -> Self {
+        self.set_interceptor(SharedInterceptor::new(interceptor));
+        self
+    }
+
+    #[doc(hidden)]
+    /// Sets an [`Interceptor`] that runs at specific stages of the request execution pipeline.
+    ///
+    /// Interceptors targeted at a certain stage are executed according to the pre-defined priority.
+    /// SDK provides the default set of interceptors. An interceptor configured by this method
+    /// will run after those default interceptors.
+    ///
+    /// ## Examples
+    /// ```no_run
+    /// # #[cfg(feature = "examples")]
+    /// # fn example() {
+    /// use aws_http::user_agent::AwsUserAgent;
+    /// use aws_smithy_runtime_api::client::interceptors::{Interceptor, InterceptorContext, SharedInterceptor};
+    /// use aws_smithy_runtime_api::config_bag::ConfigBag;
+    /// use aws_types::sdk_config::{Builder, SdkConfig};
+    /// use http::header::USER_AGENT;
+    /// use http::{HeaderName, HeaderValue};
+    ///
+    /// fn override_user_agent(builder: &mut Builder) {
+    ///     #[derive(Debug)]
+    ///     pub struct TestUserAgentInterceptor;
+    ///     impl Interceptor for TestUserAgentInterceptor {
+    ///         fn modify_before_signing(
+    ///             &self,
+    ///             context: &mut InterceptorContext,
+    ///             _cfg: &mut ConfigBag,
+    ///         ) -> Result<(), aws_smithy_runtime_api::client::interceptors::BoxError> {
+    ///             let headers = context.request_mut()?.headers_mut();
+    ///             let user_agent = AwsUserAgent::for_tests();
+    ///             headers.insert(USER_AGENT, HeaderValue::try_from(user_agent.ua_header())?);
+    ///             headers.insert(
+    ///                 HeaderName::from_static("x-amz-user-agent"),
+    ///                 HeaderValue::try_from(user_agent.aws_ua_header())?,
+    ///             );
+    ///
+    ///             Ok(())
+    ///         }
+    ///     }
+    ///     builder.set_interceptor(SharedInterceptor::new(TestUserAgentInterceptor));
+    /// }
+    ///
+    /// let mut builder = SdkConfig::builder();
+    /// override_user_agent(&mut builder);
+    /// let sdk_config = builder.build();
+    /// # }
+    /// ```
+    pub fn set_interceptor(&mut self, interceptor: SharedInterceptor) -> &mut Self {
```

Comment:
Seems like this should be called `add_interceptor`, and there should be a separate `set_interceptors` that takes an entire `Vec<SharedInterceptor>`

Reply:
I like that naming distinction. Interestingly, I originally had a plural method in a [previously closed PR](https://github.com/awslabs/smithy-rs/pull/2480/files) (although, that one was for runtime plugins and not for interceptors), and in this PR I decided to not to have the plural method `set_interceptors` because once `SdkConfig` and a service config are backed by `ConfigBag`, we no longer need to call `.set_interceptors` on `SdkConfig` to transfer the field `interceptors` from `SdkConfig` to the service config (we can just internally close the `ConfigBag`).

That said, I suppose that we could still use the plural method as a convenience method even after `SdkConfig` and a service config are backed by `ConfigBag`. Will incorporate the feedback accordingly.

Reply:
Addressed in ba26b98.